### PR TITLE
build: add ci

### LINKED ISSUE #9: add a basic ci pipeline with github actions
* linting, typechecking, testing and building both client and server

### PR DESCRIPTION
**Pull request by metehancelik**
closes #9

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  lint:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v3
+    - name: Use Node.js
+      uses: actions/setup-node@v3
+      with:
+        node-version: '18.x'
+        cache: 'npm'
```
Comment (demirtasdurmus) [inline .github/workflows/ci.yaml]: This line wont work properly unless you provide a `cache-dependency-path`. You can skip it for now as we will implement a more robust caching strategy later.

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
+name: Node.js CI
+
+on:
+  pull_request:
+    branches: [ "develop" ]
+
+jobs:
```
Comment (demirtasdurmus) [inline .github/workflows/ci.yaml:7]: These jobs will be running in seperate containers and you seem not to have connected them in a logical order. Either put all of them in one single job, or make the jobs dependent on each other in a logical way. I mean if the linting fails, there is no benefit in making the testing continue, right?

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    steps:
+    - uses: actions/checkout@v3
+    - name: Use Node.js
+      uses: actions/setup-node@v3
+      with:
+        node-version: '18.x'
+        cache: 'npm'
+    - run: npm install nx
+    - run: npm install
+    - run: cd apps/client && tsc --noEmit --esModuleInterop && cd ../server && tsc --noEmit --esModuleInterop
```
Comment (demirtasdurmus) [inline .github/workflows/ci.yaml]: You can run this action with either `nx` cli or add npm scripts to the top level.

File: package.json [hunk 4]
```diff
@@ -4,6 +4,9 @@
   "license": "MIT",
   "scripts": {
     "dev": "npx nx run-many --target=dev --projects=server,client",
+    "lint": "npx nx run-many --target=lint --projects=server,client",
+    "test": "npx nx run-many --target=test --projects=server,client",
+    "preinstall": "npm i nx",
```
Comment (demirtasdurmus) [inline package.json]: This caused problem on me. When I run `npm i`, both package.json and lock file has been changed. I dont know the reason, but if it is the preinstall, you can either solve this problem or remove preinstall.

**Review by demirtasdurmus — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (metehancelik) [reply package.json:9]: i guess i will remove

File: .github/workflows/ci.yaml (hunk 2 above)
Comment (metehancelik) [reply .github/workflows/ci.yaml:7]: it should be outdated as well

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  lint-test-type_check:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
+      - name: Use Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: "18.x"
+          cache: "npm"
```
Comment (demirtasdurmus) [inline .github/workflows/ci.yaml]: You can remove this cache field as well as we explained it.

**Review by demirtasdurmus — APPROVED** [review]
One little issue but I will approve it anyway. Its basically the cache field of the setup-node action.

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (metehancelik) [reply .github/workflows/ci.yaml]: removed

**Pull request merged**